Implementation of first version

S-Dafarra opened the pull request:
This PR implements the first version of the ``sDiff`` tool. This tool mimics automatic differentiation for expressions where known blocks are involved.

The code is built around Eigen matrices, uses templates extensively and is header only.

A proper readme is missing.

I am open to suggestions for a fancier name.

FYI
@DanielePucci, @traversaro, @GiulioRomualdi, @diegoferigo

## TODO
- [ ] README
- [x] Change name
- [x] CI

S-Dafarra committed:
First skeleton with basic operations.

S-Dafarra committed:
Handling assignements to constant. Defined skeleton for variables.

S-Dafarra committed:
Added product operation.

S-Dafarra committed:
Expressions can be set equal to themselves.

S-Dafarra committed:
Improvements on the behavior with scalars, especially on multiplication.

Gosh this was painful.

S-Dafarra committed:
Better handling of summation too.

S-Dafarra committed:
Added subtraction.

S-Dafarra committed:
Removed the definition of is_variable.

S-Dafarra committed:
Implemented move assignement and casting.

S-Dafarra committed:
Some reordering and first skeleton for derivatives.

S-Dafarra committed:
Added evaluable to extract a row and fixed casting.

S-Dafarra committed:
Refactory of VariableEvaluable.

Now it inherits from Evaluable.

S-Dafarra committed:
Moved Expression implementation to a different file.

This allows to output an expression as derivative.

S-Dafarra committed:
Moved implementation of utilities functions.

S-Dafarra committed:
Expression now depends only on the forward declarations.

S-Dafarra committed:
Intermediate modifications to use Expressions as inputs to the OperatorsEvaluables.

Missing row and element evaluables.

S-Dafarra committed:
Completed modifications to use Expressions inside the OperatorsEvaluables.

S-Dafarra committed:
Make sure that the namespace is always specified.

Commented the derivative parts to test them one by one.

S-Dafarra committed:
Added derivatives. Missing the product deerivative.

S-Dafarra committed:
Implemented product derivative.

S-Dafarra committed:
Added some documentation.

S-Dafarra committed:
Added helper method to avoid extra computation.

It checks whether an evaluable is dependent on a variable.

traversaro commented:
Cool. What's the plan? It will be a dependency of iDynTree? Or of the dynamic planner code?

traversaro commented:
Do you want to move the repo to `robotology-playground` ?

traversaro commented on CMakeLists.txt:
```diff
@@ -0,0 +1,101 @@
+install(EXPORT sDiff
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/sDiff)
+
+configure_package_config_file(${CMAKE_SOURCE_DIR}/cmake/sDiffConfig.cmake.in
```
Avoid to use `CMAKE_SOURCE_DIR` and `CMAKE_BINARY_DIR`, see https://github.com/robotology/ycm/issues/162#issuecomment-411753919 for more info.

traversaro commented on CMakeLists.txt:
```diff
@@ -0,0 +1,101 @@
+
+find_package(Eigen3 REQUIRED)
+
+include_directories(SYSTEM ${EIGEN3_INCLUDE_DIR})
```
Use `target_include_directories(sDiff PUBLIC ${EIGEN3_INCLUDE_DIR})` so that the usage is correctly propagated to the users of `sDiff`.

S-Dafarra commented:
The plan would be to test it first inside the planner, also to check whether it is already enough for the use cases I need. Afterwards I was thinking to move it in one of our organizations. 
One idea then could be to use it to simplify the definition of custom optimal control problems.

traversaro commented on cmake/sDiffConfig.cmake.in:
```diff
@@ -0,0 +1,8 @@
+set(sDiff_VERSION @sDiff_VERSION@)
+
+find_package(Eigen3 REQUIRED)
```
You can use `find_dependency` here, see  https://cmake.org/cmake/help/v3.5/module/CMakeFindDependencyMacro.html .

traversaro reviewed (changes requested):
Minor cmake comments.

traversaro commented:
> I am open to suggestions for a fancier name.

In view of future move to `robotology`, I suggest to use a repo name with just lowercase characters and eventually a `-` to separate words.

traversaro reviewed (changes requested):
Comments

traversaro commented:
I would set up the CI with at least Visual Studio as soon as possible, such template heavy code is typically a good way to encover bugs in compilers. : )

S-Dafarra commented:
> I would set up the CI with at least Visual Studio as soon as possible, such template heavy code is typically a good way to encover bugs in compilers. : )

Definitely agree. Do you have any entry point for this?

S-Dafarra committed:
CMake modifications after @traversaro review.

traversaro commented:
I am a bit confused by all the method that take in input a `std::shared_ptr` . A const reference would not be enough in those cases?

traversaro commented:
> Definitely agree. Do you have any entry point for this?

As for anything related to how to develop C++ libraries, https://github.com/robotology/how-to-export-cpp-library/blob/master/.appveyor.yml .

traversaro commented:
> I am open to suggestions for a fancier name.

`semi-auto-diff` ? `sdiff` has a lot of collisions as a name ( https://www.google.com/search?ei=5TZPXNSDFYeRsAf7g5WgCg&q=sdiff+github&oq=sdiff+github&gs_l=psy-ab.3..0i13l3j0i13i30l6j0i13i5i30.5671.6388..6560...0.0..0.149.912.0j7......0....1..gws-wiz.......0i71j0i20i263j0j0i67j0i22i30j0i22i10i30._RitgoD9Fqk ).

S-Dafarra commented:
> I am a bit confused by all the method that take in input a std::shared_ptr . A const reference would not be enough in those cases?

The two methods that need this are ``getColumnDerivative`` and ``isDependentFrom``. The use case is the one of this line:
https://github.com/S-Dafarra/sDiff/blob/6e94844bcb89242024cc94cfb79d7831f1201417/tests/DerivativeTest.cpp#L42

where ``x`` is defined as a ``Variable``. This class carries also a vector used to store the values which I don't want to move around. Also, ``Variable`` is a ``typedef`` of a more complex type (https://github.com/S-Dafarra/sDiff/blob/6e94844bcb/include/sDiff/ForwardDeclarations.h#L177). By passing the reference I would have been bound to make those method templatized (is it a word?), preventing me to have them virtual here (https://github.com/S-Dafarra/sDiff/blob/6e94844bcb/include/sDiff/Evaluable.h#L171-L184).

S-Dafarra committed:
Moved find_dependency to the config.cmake file.

S-Dafarra commented:
> semi-auto-diff ? sdiff has a lot of collisions as a name

I would go for something as short as possible :grin:. Possible keywords I thought of were:
- semi-automatic
- block-wise
- differentiation
- column-wise
- evaluable

S-Dafarra committed:
Implemented block operator.

traversaro commented:
Just `blockwise` it is quite cool and not a lot of collisions (hoping that it does not generate confusion with `BlockFactory`).

S-Dafarra commented:
> Just blockwise it is quite cool and not a lot of collisions (hoping that it does not generate confusion with BlockFactory).

Agree! I would wait till the very last to perform this modification anyway.

S-Dafarra committed:
Added skew operator and test with rotation matrix.

S-Dafarra committed:
Added test on rotation derivative.

S-Dafarra committed:
Added constructor by rows.

S-Dafarra committed:
Added constructor by columns.

S-Dafarra committed:
Added transpose.

S-Dafarra committed:
Considering also the case in which we multiply times an unitary matrix.

S-Dafarra committed:
Added pow and division operators.

S-Dafarra committed:
Moved accessors evaluables to a separate file.

S-Dafarra committed:
Using not normalized quaternion in Rotation test.

S-Dafarra committed:
Added test on double derivative of rotation.

S-Dafarra commented:
List of bad names :sweat_smile: :
- SAD (**S**emi-**A**utomatic **D**ifferentiation)
- BAD (**B**lockwise **A**utomatic **D**ifferentiation)
- SAUDI (**S**emi-**AU**tomatic **DI**fferentiation)
- AD-BLOCK (**A**utomatic **D**ifferentiation - **BLOCK**wise)

DanielePucci commented:
@S-Dafarra 
It would be nice to add also the meaning of the acronyms, just to know what you have in mind ;-)

In any case, I vote for SAD 😄

S-Dafarra commented:
> It would be nice to add also the meaning of the acronyms, just to know what you have in mind ;-)

Comment updated :grin:

S-Dafarra commented:
Some interesting acronyms obtained by putting the [keywords above](https://github.com/S-Dafarra/sDiff/pull/1#issuecomment-458223070) in http://acronymcreator.net/ace.py:
- BLADE | BLock Automatic Differentiation of Evaluables
- CODE | COlumnwise Differentiation of Evaluables
- BLOCKADE | BLOCK Automatic Differentiation of Evaluables
- COLD | COLumnwise Differentiation
- BESIDE | BlockwisE SemIautomatic Differentiation Evaluable

traversaro commented:
I like `BloackAD` or `SemiAD`.

S-Dafarra commented:
Ok I think I'll drop any reference to automatic differentiation in the name. It just does not look good to me to put this piece of code in the automatic differentiation framework. As a consequence also the idea of an acronym is dropped. At this point I would go for a single word name.

I think I will go for **``levi``**, from [Tullio Levi-Civita](https://en.wikipedia.org/wiki/Tullio_Levi-Civita). He was an italian mathematician who invented the tensor calculus. Clearly this library is not about tensor calculus, but inside I have used the [Levi-Civita symbol](https://en.wikipedia.org/wiki/Levi-Civita_symbol). In particular the [Levi-Civita symbol in three dimensions corresponds](https://en.wikipedia.org/wiki/Levi-Civita_symbol#Three_dimensions) to the column derivative of the skew symmetric matrix obtained from a 3D vector. 

It is also quite short and catchy as a name :grin: 

What do you think?

traversaro commented:
Ok for me (should not have a lot of collision, see https://repology.org/metapackages/?search=levi and https://github.com/topics/levi).

DanielePucci commented:
@S-Dafarra 
For me it is fine, though **`levi`** is not the acronym of anything and it makes difficult to remember what you do from the name

S-Dafarra committed:
Changed name to levi.

S-Dafarra commented:
I know, but none of the acronyms was satisfactory, all a bit "sad" :grin:

S-Dafarra committed:
Moved some template declarations in a separate file.

S-Dafarra committed:
Added CI.

S-Dafarra committed:
Attempt to fix appveyor.

traversaro commented on CMakeLists.txt:
```diff
@@ -0,0 +1,87 @@
+        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+
+install(EXPORT levi
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/levi)
```
I strongly suggest to add a NAMESPACE with the same name of the CMake package (see https://github.com/robotology/how-to-export-cpp-library/blob/782e9b1b13867686463b2a8a5b2b9d67aa176099/cmake/InstallBasicPackageFiles.cmake#L673) so that your target will be usable as `levi::levi`.

S-Dafarra replied:
Is it enough to add ``NAMESPACE levi::`` in the ``install`` command?

traversaro replied:
Yes. 

I also suggest to add a `levi::levi` ALIAS to the target of the library, see https://github.com/robotology/how-to-export-cpp-library/blob/e23e399fce102ba8f87cd5269797513000b99794/src/LibTemplateCMake/CMakeLists.txt#L62 . This simplifies the process of consuming the library with [`FetchContent`+`add_subdirectory`](https://cmake.org/cmake/help/v3.11/module/FetchContent.html), that is quite fashionable these in CMake world.

S-Dafarra committed:
Yet another attempt for appveyor.

S-Dafarra committed:
Avoid to use std::min in template.

S-Dafarra committed:
Fix of windows compilation bugs.

Using the full template for Eigen matrices.
Different initialization of the vector of expressions in the advanced constructors.

S-Dafarra commented:
Apparently I still have two issues with Windows:
- VS2015 does not seem to be fully compatible with SFINAE.
- Numeric test on the ``RotationTest`` have hugely different results (compared to ubuntu)

S-Dafarra commented:
> Numeric test on the RotationTest have hugely different results (compared to ubuntu)

Apparently this was due to these two lines (I will commit the modification later)
https://github.com/S-Dafarra/levi/blob/2c88ebf2a930fed0b23e7b454ad4757ef00eaafb/tests/RotationTest.cpp#L116-L117

which are purely ``Eigen`` methods 😕

traversaro commented:
I guess it is ok to just support vs2017

S-Dafarra commented:
Don't you think it will limit the possibilities to use it in other projects?

S-Dafarra committed:
Silenced warning about unused parameters.

S-Dafarra committed:
Fixed rotation test on windows.

traversaro commented:
> Don't you think it will limit the possibilities to use it in other projects?

At least in "our" robotology projects, I  do not think so. We have a sort of custom of just supporting the latest two versions of Visual Studio, and Visual Studio 2019 is coming out soon. 

Furthermore, one of the main reason in the past for supporting old versions of VS was that if you had  to work with a driver library available in binary form for a given version of VS, you typically had to build all your software with that version of Visual Studio. But from VS2015, the ABI and CRT libraries are forward-compatible, meaning that you can use a library compiled with VS2015 in a executable compiled with VS2017+, see https://docs.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017?view=vs-2017 . This, coupled with the fact that on Windows you are not tied with a fixed compiler of your system to keep compatibility with system libraries, means that is relativy easy for Windows developers to just use the latest Visual Studio version.

traversaro commented:
> Numeric test on the RotationTest have hugely different results (compared to ubuntu)

You are using a quite old Eigen (3-3beta2) version in AppVeyor tests. Are you sure that this strange results are not related to that? I used it in iDynTree for Xenial consistency, but perhaps there are some Windows-related bugs fixed in later versions.

S-Dafarra commented:
Actually I had the same issue also on my Windows pc where I installed the 3.3.7.

S-Dafarra committed:
Suppression of other unused variables warnings.

S-Dafarra committed:
Dropping VS2015.

S-Dafarra committed:
Using variables and namespace in the CMakeLists file.

S-Dafarra committed:
Resize is now virtual to allow null and identity evaluable to be reinitialized correctly.

S-Dafarra committed:
Added mechanism to reuse buffers when possible.

S-Dafarra committed:
Avoid ambiguity in constructors and copy operators.

S-Dafarra committed:
Attempt to reduce number of evaluations.

S-Dafarra committed:
Removed dynamic cast of pointer.

S-Dafarra committed:
Few improvements to a better reuse of buffers.

A simple flag is used to check if the evaluate method has been called at least once.
The identity derivative in the EvaluableVariable is a parameter of the class.
Updated gitignore to remove windows files.

S-Dafarra committed:
AccessorEvaluables inherit from a common base class.

S-Dafarra committed:
Operators now inherit from a common class.

It could be either Unary or BinaryOperator.

S-Dafarra committed:
Reuse of expressions when computing derivatives.

S-Dafarra committed:
Reusing expressions also in binary operators and advanced constructors.

S-Dafarra committed:
Adding valgrind tests.

S-Dafarra committed:
ForwardDeclarations can be used to forward declare levi types.

S-Dafarra committed:
The evaluation register uses a recycle bin for dead IDs.

This is to avoid that repeated local copies of an expression will make the register grow indefinitely.

S-Dafarra committed:
Added other common expressions in forwardDeclarations.

S-Dafarra committed:
Column expressions can be used as a variable.

S-Dafarra committed:
Using const where possible.

For some methods it is not possible to use const. This has to be investigated.

S-Dafarra committed:
Using caching of derivatives directly in evaluable.

S-Dafarra committed:
Allowing to clear the cache of derivatives.

S-Dafarra committed:
Added Horzcat and Vertcat.

S-Dafarra committed:
Added check to avoid casting and expression to a variable.

S-Dafarra committed:
Defined fixed size expressions.

S-Dafarra committed:
Using _ENABLE_EXTENDED_ALIGNED_STORAGE on windows.

S-Dafarra committed:
Moved _ENABLE_EXTENDED_ALIGNED_STORAGE in levi.h

S-Dafarra committed:
Using cpp17 on WIndows to workaround the alignement issues on Eigen.

S-Dafarra committed:
Minor improvements.

- Evaluable checks if it is dependent from the variable before asking for new derivatives
- Using lazy assign
- Thanks to a dummy int in some constructors, all the operators are now const.
- Using std::fill to reinitialize the evaluation register.

S-Dafarra committed:
The expression can detect if it is null. Row, col and element can be retrieved from the evaluable.

If the expression has suitable dimensions, return *this when calling row, col or operator().

S-Dafarra committed:
Operators avoid to create new evaluables in case of null operand.

Constructor by row/col can return their row/col.

S-Dafarra committed:
Returning row, col and element where possible.

S-Dafarra committed:
Using master version of Eigen in appveyor.

S-Dafarra committed:
Trying version 3.3 of Eigen in appveyor.

S-Dafarra committed:
Updated FindEigen3.

S-Dafarra committed:
Fixed Eigen path in appveyor.

Thanks @traversaro

S-Dafarra committed:
Fixing again Eigen3 path in appveyor.

S-Dafarra committed:
Product, pow and division override the row/col/element methods.

S-Dafarra committed:
Skew, transpose and block implement row, col, element.

S-Dafarra committed:
Using /bigobj in apveyor.

S-Dafarra committed:
Avoid to perform the product when one of the operand is the identity.

Fixed bug in the computation of the derivative of pow.

S-Dafarra committed:
Added SingleElementMatrix and TwoElementsMatrix evaluables.

These are useful to specify rows and cols of identities and levi-civita symbols.

S-Dafarra committed:
Defining out-of-line virtual destructors.

S-Dafarra committed:
Avoid calling isNew too many times.

S-Dafarra committed:
Fixed windows compilation.

S-Dafarra committed:
Using method to check if dependencies are new.

S-Dafarra committed:
Restored bool when calling evaluate().

ComposeByRows and ByCols can be accessed also outside of an expression.

S-Dafarra committed:
Assert on dimension of derivatives.

S-Dafarra committed:
Added helper class to deal with dependencies.

S-Dafarra committed:
Fixed DependenciesSet on windows.

S-Dafarra committed:
Added enum to save which particular evaluable an expression is pointing to.

S-Dafarra committed:
Using info structure to preserve tree structure.

S-Dafarra committed:
Added squeeze evaluable.

S-Dafarra committed:
Bugfixes in SqueezeEvaluable.

S-Dafarra committed:
Fixed stupid bug on RotationTest.

S-Dafarra committed:
Added EvaluableInfo to Evaluable class.

S-Dafarra committed:
Removed EvaluableInfo from Expression.

Apparently having this struct into the expression was slowing down a lot (factor 2) all the evaluations.

S-Dafarra committed:
Using full template definition when adding CastEvaluables as a friend.

This was causing a failure in windows compilation.

Using lazyAssign in SqueezeEvaluable.

S-Dafarra committed:
Keeping scalars always on the lhs when multiplying two evaluables.

S-Dafarra committed:
Added unary + operator and ==.

S-Dafarra committed:
Condensing subexpressions when summing expressions.

S-Dafarra committed:
Added simplifications when one of the operands just invert signs.

S-Dafarra committed:
The value in the levi triplet is an expression.

This allows to return a scalar identity (1.0) when extracting a particular element of the levi-civita elements.

S-Dafarra committed:
Removed few unwanted casts.

S-Dafarra committed:
Added first draft of mechanism to autogenerate expressions.

S-Dafarra committed:
Avoid using ps to compile cppfs in appveyor.

S-Dafarra committed:
Fixed generations of expressions by dealing with the case in which an expression is scalar.

Also, transpose() return *this when rows=cols=1.

S-Dafarra committed:
Running cmake on the autogenerated expressions.

Few cleanups.

S-Dafarra committed:
Using zupply (a partially cleaned version) to deal with creation of directories.

S-Dafarra committed:
Fixed windows automatic compilation.

The files were created with strange names due to conversion from string to wstring. Using normal fstream to fill files once the directory is created.

S-Dafarra committed:
Using shlibpp to dynamically link generated evaluables.

S-Dafarra committed:
Fixed memory leak when using factory.

S-Dafarra committed:
Autogeneration completed and moved to a separate file.

Using a different strategy in cmake to make sure that the library is in the lib folder. Checking the existance of the library before trying to load it.

S-Dafarra committed:
Using a helper class to expand the tree. Added a test to benchmark evaluation performances.

Committing AutogeneratedEvaluable.

S-Dafarra committed:
For the operators == returns true even if the hash is different.

S-Dafarra committed:
In the autogenerated evaluable the doubles are saved in a local variable.

Fixed bug in the CompiledRotationTest. The computation of the derivatives was wrong due to the use of asVariable.

S-Dafarra committed:
Precomputing common subexpressions in the autogenerated evaluable.

S-Dafarra committed:
Removing unnecessary common subexpressions.

S-Dafarra committed:
Disabled subexpressions shall not count when checking is a subexpression is useful.

S-Dafarra committed:
Turned TreeExpander into a function.

First step to have compilation of multiple expressions.

S-Dafarra committed:
Moved nearly all the logic of the JIT in a helper class to ease the compilation of multiple expressions.

S-Dafarra committed:
Added possibility to compile and execute multiple expressions.

S-Dafarra committed:
Using a simpler input for the multiple expression compilation.

S-Dafarra committed:
Added some typedefs and aliases to simplify the use of CompileMultipleExpressions.

S-Dafarra committed:
Avoiding to save element type in helper variables when generating code.

Added some debug prints.

S-Dafarra committed:
First attempt to fix appveyor.

S-Dafarra committed:
Added possibility to squeeze multiple expressions.

S-Dafarra committed:
Speed up of detection of common subexpressions.

Allowing to specify the name of generics, helpers and commons.

S-Dafarra committed:
TreeExpander can avoid to allocate memory.

S-Dafarra committed:
Helpers variables are clean as well. Using Ninja generator if available.

S-Dafarra committed:
Constants horzcat and vertcat are also considered when squeezing (but not when compiling)

S-Dafarra committed:
The tree is expanded avoiding duplicating branches. This increased a bit the time to compute the squeeze evaluable.

Improved definitionof the operator == and !=.

S-Dafarra committed:
Using a vector of int for the evaluation register.

This remove the need of setting to false all the entries of the register in case of new values.

S-Dafarra committed:
Refactored mechanism of isNew and isDependentFrom.

The concept of Registrar is introduced, whose aim is to avoid having to go through all the expression tree to know if the expression is new. Indeed, only few evaluables can change their values.
In particular, these are variables and mutables. The latter are similar to constants with the difference that they have defined the operator equal. All the methods isNew and isDependent from
have been substituted by the method addDependencies where the actual dependencies are specified.

S-Dafarra committed:
Clearing of derivatives caches goes all the way down of the expression tree.

S-Dafarra committed:
Added possibility to get an independent variable from an expression.

S-Dafarra committed:
Allowing to define names of scalar Constants and Mutables.

S-Dafarra committed:
Protecting Evaluable in case of multithreaded evaluation.

S-Dafarra committed:
Added option to enable CompiledRotation test.

S-Dafarra committed:
Simplified the product derivative.

S-Dafarra committed:
Trying to evaluate products from smaller to bigger factors to minimize computational cost.

S-Dafarra committed:
Specifying CMAKE_GENERATOR for blockfactory in appveyor.

S-Dafarra committed:
Using powershell when installing blockfactory in appveyor.

S-Dafarra committed:
Attempt to fix appveyor.

S-Dafarra committed:
Dependencies use raw pointers to Registrars. Added Assignable.

Having raw pointers allows to store (this) as dependency. In turn this allowed to define the Assignable object, where the operator = is defined, while avoiding to override the getDependencies method.
In turn, the static check for the = operator is easier. This seemed to be causing issues in VCC and clang.

S-Dafarra committed:
Forgot to push Assignable file.

S-Dafarra committed:
Avoiding recursion when calling isNew or isDependentFrom when (this) is a depndency.

S-Dafarra committed:
First row/col of ConstructorsByRows/Cols was not take into account in the dependencies.

S-Dafarra committed:
Removed useless casts when computing the product cost.

S-Dafarra commented:
This PR has definitively gone out of control. Sorry for the spam. Merging as is, facing the remaining problems in other issues.
